### PR TITLE
RFR: Add Admin & Authenticated route protection

### LINKED ISSUE #157: FE: Admin & Logged in route protection
**Is your feature request related to a problem? Please describe.**

Right now, any user can access the admin pages. 

**Describe the solution you'd like**

We want only the admins to be able to do this.

### PR DESCRIPTION
**Pull request by XuanLi-Leong**
## Status: 

:rocket: Ready

## Description
No joke, learning about this react children, higher order components and container components was fun

Related issues: Closes #157

### CODE REVIEW

File: frontend/src/App.js [hunk 1]
```diff
@@ -21,7 +22,34 @@ import AccountManagePage from './pages/AccountManagePage';
 import BadgesPage from './pages/BadgesPage';
 
 import './styles/App.scss';
-import AdminEditBookPage from './pages/admin/AdminEditBookPage';
+import AdminProtection from './pages/protection/AdminProtection';
+import AuthenticatedProtection from './pages/protection/AuthenticatedProtection';
+
+function withAdminProtection(WrappedComponent) {
```
Comment (XuanLi-Leong) [inline frontend/src/App.js]: Higher-Order Component stuff

File: frontend/src/pages/protection/AdminProtection.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    if (!isLoggedIn) {
+      return <Redirect push to="/login" />;
+    }
+
+    if (!isAdmin) {
+      return <Redirect to="/" />;
+    }
+
+    return (
+      <div>{React.cloneElement(this.props.children, { ...this.props })}</div>
```
Comment (XuanLi-Leong) [inline frontend/src/pages/protection/AdminProtection.js]: Basically a container component

File: frontend/src/pages/protection/AdminProtection.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    this.state = {
+      isLoggedIn: undefined,
+      isAdmin: undefined
+    };
+    this.checkIsAdmin();
+  }
+
+  checkIsAdmin = async () => {
+    const { success, result, message } = await getUserData();
+    console.log(result.is_admin);
```
Comment (josh-byster) [inline frontend/src/pages/protection/AdminProtection.js]: Might be able to remove this for prod?

Comment (josh-byster) [inline frontend/src/pages/protection/AdminProtection.js]: Small change: could maybe remove `console.log` for prod?

File: frontend/src/pages/protection/AuthenticatedProtection.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    if (isLoggedIn === undefined) {
+      return null;
+    }
+
+    if (!isLoggedIn) {
+      return <Redirect push to="/login" />;
+    }
+
+    return (
+      <div>{React.cloneElement(this.props.children, { ...this.props })}</div>
```
Comment (josh-byster) [inline frontend/src/pages/protection/AuthenticatedProtection.js]: Wow, awesome stuff! 🚀

**Pull request merged**